File: baseline.ts
import axios from 'axios';
import { GoogleGenAI } from '@google/genai';

const API_URL = "http://localhost:3000";
const GEMINI_API_KEY = process.env.GEMINI_API_KEY || "";

const ai = new GoogleGenAI({ apiKey: GEMINI_API_KEY });

async function runTask(taskId: string) {
  console.log(`\n--- Running Task ${taskId} ---`);
  const resetRes = await axios.post(`${API_URL}/reset?task_id=${taskId}`);
  let obs = resetRes.data;
  
  let done = false;
  let totalReward = 0;
  
  while (!done) {
    const prompt = `
      You are an office assistant. 
      Current Observation: ${JSON.stringify(obs)}
      Available Actions: classify_email, schedule_meeting, respond_ticket, clean_data.
      
      Choose one action and return it as JSON in the format:
      {"type": "action_type", "parameters": {...}}
      Only return the JSON.
    `;

    let action;
    if (!GEMINI_API_KEY || GEMINI_API_KEY === "MY_GEMINI_API_KEY") {
      // Heuristic baseline if no key
      if (taskId === "1") {
        const unclassified = obs.emails.find((e: any) => !e.category);
        if (unclassified) {
          action = { type: "classify_email", parameters: { email_id: unclassified.id, category: "urgent" } };
        } else {
          action = { type: "clean_data", parameters: {} };
        }
      } else if (taskId === "2") {
        action = { type: "schedule_meeting", parameters: { title: "HR Meeting", start_time: "10:00", end_time: "11:00" } };
      } else if (taskId === "3") {
        const openTicket = obs.tickets.find((t: any) => t.status === "open");
        if (openTicket) {
          action = { type: "respond_ticket", parameters: { ticket_id: openTicket.id, response: "Fixed" } };
        } else {
          action = { type: "clean_data", parameters: {} };
        }
      } else {
        action = { type: "clean_data", parameters: {} };
      }
    } else {
      const response = await ai.models.generateContent({
        model: "gemini-3-flash-preview",
        contents: prompt,
        config: { responseMimeType: "application/json" }
      });
      action = JSON.parse(response.text || "{}");
    }

    console.log(`Action: ${JSON.stringify(action)}`);
    await new Promise(r => setTimeout(r, 1500)); // Artificial delay to watch the UI!
    const stepRes = await axios.post(`${API_URL}/step`, action);
    const stepData = stepRes.data;
    obs = stepData.observation;
    totalReward += stepData.reward;
    done = stepData.done;
    
    if (stepData.info?.reason) {
      console.log(`Result: ${stepData.info.reason}`);
    }
  }

  const gradeRes = await axios.get(`${API_URL}/grader`);
  console.log(`Final Score: ${gradeRes.data.score}`);
  return gradeRes.data.score;
}

async function main() {
  const scores: Record<string, number> = {};
  for (const tid of ["1", "2", "3"]) {
    try {
      scores[tid] = await runTask(tid);
    } catch (e) {
      console.error(`Error running task ${tid}:`, e);
    }
  }
  
  console.log("\n--- Summary ---");
  Object.entries(scores).forEach(([tid, score]) => {
    console.log(`Task ${tid}: ${score}`);
  });
}

main();

File: baseline.py
import os
import time
import json
import requests
from openai import OpenAI

API_URL = os.environ.get("OPENENV_API_URL", "http://localhost:8000")

try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass

OPENAI_API_KEY = os.environ.get("OPENAI_API_KEY")
if not OPENAI_API_KEY:
    print("Error: OPENAI_API_KEY environment variable not set. Please set it in your .env file.")
    exit(1)

client = OpenAI(api_key=OPENAI_API_KEY)

USE_HEURISTIC_FALLBACK = False

def run_task(task_name: str):
    global USE_HEURISTIC_FALLBACK
    print(f"\n--- Running Baseline for Task: {task_name} ---")
    
    res = requests.post(f"{API_URL}/reset", json={"task_name": task_name})
    if res.status_code != 200:
        print(f"Failed to reset: {res.text}")
        return 0.0
    
    state_data = res.json()
    state = state_data.get("observation", state_data)
    
    system_prompt = """
    You are an AI assistant in an office environment. 
    You manage emails, a calendar, support tickets, and pending tasks.
    Perform the pending tasks by choosing the appropriate action.
    Valid actions (choose exactly ONE and output AS JSON ONLY): 
    - {"action_type": "classify_email", "email_id": "str", "target_folder": "str"}
    - {"action_type": "schedule_meeting", "time": "str", "title": "str", "participants": ["str"]}
    - {"action_type": "respond_ticket", "ticket_id": "str", "response": "str"}
    - {"action_type": "clean_data", "target_system": "str", "record_id": "str", "action": "str"}
    
    Output ONLY a valid JSON object. No other text, markdown, or chat formatting.
    """

    max_steps = 5
    for step_num in range(max_steps):
        print(f"Step {step_num + 1}...")
        
        state_str = json.dumps(state, indent=2)
        
        action = None
        from heuristic_agent import get_heuristic_action
        
        if USE_HEURISTIC_FALLBACK:
            action = get_heuristic_action(state, task_name)
            time.sleep(1.5)  # Visual pacing for the fallback UI
            print(f"Action (Heuristic Fallback): {action.get('action_type', 'unknown')}")
        else:
            for attempt in range(5):
                try:
                    response = client.chat.completions.create(
                        model="gpt-4o-mini",
                        messages=[
                            {"role": "system", "content": system_prompt},
                            {"role": "user", "content": f"Current State:\n{state_str}\n\nWhat is your next action?"}
                        ],
                        response_format={"type": "json_object"}
                    )
                    action_text = response.choices[0].message.content.strip()
                    import re
                    match = re.search(r'(\{.*\})', action_text, re.DOTALL)
                    if match:
                        action_text = match.group(1)
                        
                    action = json.loads(action_text)
                    break
                except Exception as e:
                    print(f"Error calling OpenAI API: {e}")
                    if "insufficient_quota" in str(e) or "429" in str(e) or "401" in str(e):
                        print("API Quota Reached! Switching seamlessly to Rule-Based Heuristic Fallback...")
                        USE_HEURISTIC_FALLBACK = True
                        action = get_heuristic_action(state, task_name)
                        break
                    
                    if attempt < 4:
                        print("Retrying after 5 seconds...")
                        time.sleep(5)
                    else:
                        print("Max retries reached. Switching to Heuristic Fallback...")
                        USE_HEURISTIC_FALLBACK = True
                        action = get_heuristic_action(state, task_name)
                        break
        
        if not action:
            break
            
        try:
            print(f"Agent Action: {action.get('action_type', 'Unknown')}")
            
            step_res = requests.post(f"{API_URL}/step", json={"action": action})
            if step_res.status_code != 200:
                print(f"Action failed: {step_res.text}")
                break
                
            reward_data = step_res.json()
            is_done = reward_data.get("is_done", False)
            if is_done:
                print("Task marked as done by environment.")
                break
                
            new_state_data = requests.get(f"{API_URL}/state").json()
            state = new_state_data.get("observation", new_state_data)
            
        except Exception as e:
            print(f"Error during baseline run: {e}")
            break
            
    score_res = requests.get(f"{API_URL}/grader")
    if score_res.status_code == 200:
        score = score_res.json().get("score", 0.0)
        print(f"Final Score for {task_name}: {score}")
        return score
    else:
        print(f"Failed to get score: {score_res.text}")
        return 0.0

if __name__ == "__main__":
    try:
        requests.get(f"{API_URL}/tasks")
    except requests.exceptions.ConnectionError:
        print(f"Error: API server is not running at {API_URL}.")
        print("Start it with: python -m uvicorn main:app --port 8000")
        exit(1)

    tasks_to_run = ["email_triage", "meeting_scheduling", "customer_support"]
    scores = {}
    
    for task in tasks_to_run:
        scores[task] = run_task(task)
        
    print("\n=== Baseline Results ===")
    for t, s in scores.items():
        print(f"{t}: {s}")
